// passport-yahoo-token/lib/index.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _passportOauth = require('passport-oauth');

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * `YahooTokenStrategy` constructor.
 * The Yahoo authentication strategy authenticates requests by delegating to Yahoo using the OAuth 2.0 protocol.
 * Applications must supply a `verify` callback which accepts an accessToken, refreshToken, profile, and callback.
 * Callback supplying a `user`, which should be set to `false` if the credentials are not valid.
 * If an exception occurred, `error` should be set.
 *
 * Options:
 * - clientID          your Yahoo application's App ID
 * - clientSecret      your Yahoo application's App Secret
 * - passReqToCallback If need, pass req to verify callback
 *
 * @param {Object} _options
 * @param {Function} _verify
 * @example
 * passport.use(new YahooTokenStrategy({
 *   clientID: '123456789',
 *   clientSecret: 'shhh-its-a-secret'
 * }), function(req, accessToken, refreshToken, profile, next) {
 *   User.findOrCreate({yahooId: profile.id}, function(error, user) {
 *     next(error, user);
 *   })
 * });
 */

var YahooTokenStrategy = function (_OAuth2Strategy) {
  _inherits(YahooTokenStrategy, _OAuth2Strategy);

  function YahooTokenStrategy(_options, _verify) {
    _classCallCheck(this, YahooTokenStrategy);

    var options = _options || {};
    var verify = _verify;

    options.authorizationURL = options.authorizationURL || 'https://api.login.yahoo.com/oauth2/request_auth';
    options.tokenURL = options.tokenURL || 'https://api.login.yahoo.com/oauth2/get_token';

    var _this = _possibleConstructorReturn(this, Object.getPrototypeOf(YahooTokenStrategy).call(this, options, verify));

    _this.name = 'yahoo-token';
    _this._accessTokenField = options.accessTokenField || 'access_token';
    _this._refreshTokenField = options.refreshTokenField || 'refresh_token';
    _this._xoauthYahooGuidField = options.xoauthYahooGuidField || 'xoauth_yahoo_guid';
    _this._profileURL = options.profileURL || 'https://social.yahooapis.com/v1/user/:xoauthYahooGuid/profile?format=json';
    _this._passReqToCallback = options.passReqToCallback;

    _this._oauth2.useAuthorizationHeaderforGET(true);
    return _this;
  }

  /**
   * Authenticate method
   * @param {Object} req Request stream
   * @param {Object} options Options object
   * @returns {*}
   */

  _createClass(YahooTokenStrategy, [{
    key: 'authenticate',
    value: function authenticate(req, options) {
      var _this2 = this;

      var accessToken = req.body && req.body[this._accessTokenField] || req.query && req.query[this._accessTokenField];
      var refreshToken = req.body && req.body[this._refreshTokenField] || req.query && req.query[this._refreshTokenField];
      var xoauthYahooGuid = req.body && req.body[this._xoauthYahooGuidField] || req.query && req.query[this._xoauthYahooGuidField];

      if (!accessToken || !xoauthYahooGuid) return this.fail({ message: 'You should provide ' + this._accessTokenField + ' and ' + this._xoauthYahooGuidField });

      this._loadUserProfile({
        accessToken: accessToken,
        xoauthYahooGuid: xoauthYahooGuid
      }, function (error, profile) {
        if (error) return _this2.error(error);

        var verified = function verified(error, user, info) {
          if (error) return _this2.error(error);
          if (!user) return _this2.fail(info);

          return _this2.success(user, info);
        };

        if (_this2._passReqToCallback) {
          _this2._verify(req, accessToken, refreshToken, profile, verified);
        } else {
          _this2._verify(accessToken, refreshToken, profile, verified);
        }
      });
    }

    /**
     * Get user profile from Yahoo
     * @param {String} accessToken Yahoo access token
     * @param {String} xoauthYahooGuid Yahoo GUID
     * @param {Function} done
     */

  }, {
    key: 'userProfile',
    value: function userProfile(accessToken, xoauthYahooGuid, done) {
      this._oauth2.get(this._profileURL.replace(':xoauthYahooGuid', xoauthYahooGuid), accessToken, function (error, body, res) {
        if (error) {
          try {
            var errorJSON = JSON.parse(error.data);
            return done(new _passportOauth.InternalOAuthError(errorJSON.error.description, error.statusCode));
          } catch (_) {
            return done(new _passportOauth.InternalOAuthError('Failed to fetch user profile', error));
          }
        }

        try {
          var json = JSON.parse(body).profile;
          json['id'] = json.guid;

          var profile = {
            provider: 'yahoo',
            id: json.id,
            displayName: [json.givenName || '', json.familyName || ''].join(' '),
            name: {
              familyName: json.familyName || '',
              givenName: json.givenName || ''
            },
            emails: [{
              value: json.emails && json.emails[0].handle || '',
              type: json.emails && json.emails[0].type || ''
            }],
            photos: [{
              value: json.image && json.image.imageUrl || ''
            }],
            _raw: body,
            _json: json
          };

          done(null, profile);
        } catch (e) {
          done(e);
        }
      });
    }

    /**
     * Load user profile
     * @param {Object} params
     * @param {Function} done
     * @private
     */

  }, {
    key: '_loadUserProfile',
    value: function _loadUserProfile(params, done) {
      return this.userProfile(params.accessToken, params.xoauthYahooGuid, done);
    }
  }]);

  return YahooTokenStrategy;
}(_passportOauth.OAuth2Strategy);

exports.default = YahooTokenStrategy;
module.exports = exports['default'];
